### PR TITLE
Implement the array length enforcer

### PR DESCRIPTION
**Pull request by aabounegm**
that makes sure all types have lengths except possible in parameters

### CODE REVIEW

File: ast/impl/array_length_enforcer.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+void ArrayLengthEnforcer::visit(IntegerType*) {}
+
+void ArrayLengthEnforcer::visit(RealType*) {}
+
+void ArrayLengthEnforcer::visit(BooleanType*) {}
+
+void ArrayLengthEnforcer::visit(ArrayType* node) {
+    if (!insideParameters && node->length == nullptr) {
+        error(node->begin, "Array size can only be left out in parameters");
```
Comment (kureduro) [inline ast/impl/array_length_enforcer.cpp:43]:
```suggestion
        error(node->begin, "array size omitted in non-signature context");
```

**Review by kureduro — APPROVED** [review]

File: ast/impl/array_length_enforcer.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        routine->accept(*this);
+    }
+}
+
+void ArrayLengthEnforcer::visit(RoutineDecl* node) {
+    insideParameters = true;
+    for (auto param : node->parameters) {
+        param->type->accept(*this);
+    }
+    insideParameters = false;
```
Comment (MefAldemisov) [inline ast/impl/array_length_enforcer.cpp:21]: Shouldn't there be a `m_` prefix for name of global variable?

Comment (aabounegm) [reply ast/impl/array_length_enforcer.cpp:21]: It is not a global variable, it is a private variable inside this class only.
Also, I didn't really understand what `m_` and `g_` mean

Comment (kureduro) [reply ast/impl/array_length_enforcer.cpp:21]: g_ stands for global.
m_ I don't know. But is looks better that a leading or trailing underscore.

**Review by MefAldemisov — APPROVED** [review]

**Pull request merged**